lib/rules/class-methods-use-this.js

```diff
@@ -125,10 +134,22 @@ module.exports = {
             "FunctionDeclaration:exit": exitFunction,
             FunctionExpression: enterFunction,
             "FunctionExpression:exit": exitFunction,
-            "PropertyDefinition > ArrowFunctionExpression.value": enterFunction,
-            "PropertyDefinition > ArrowFunctionExpression.value:exit": exitFunction,
+
+            /*
+             * Consuming marked 'this' in PropertyDefinition.
+             * ex: class foo { a = this; }
+             */
+            PropertyDefinition: () => stack.push(false),
+            "PropertyDefinition:exit": () => stack.pop(),
```

Comment:
`this` inside initializer expressions is an instance of the class (or the class itself if the field is static), but `this` inside `[]` is that of the enclosing function.

```js
/* eslint class-methods-use-this: "error" */

class C {
  doesUseThis() { // false positive
    return class { [this.foo] = 1; };
  }
}
```